library/src/main/java/com/datatorrent/lib/state/managed/AbstractManagedStateImpl.java

```diff
@@ -0,0 +1,529 @@
+
+  private transient int operatorId;
+  private transient long windowId;
+
+  private transient int windowCount;
+
+  private transient long largestRecoveryWindow;
+  protected transient boolean replay;
+
+  @NotNull
+  protected Comparator<Slice> keyComparator = new SliceComparator();
+
+  protected final transient AtomicReference<Throwable> throwable = new AtomicReference<>();
+
+  @NotNull
+  @FieldSerializer.Bind(JavaSerializer.class)
+  private Duration checkStateSizeInterval = Duration.millis(
+      DAGContext.STREAMING_WINDOW_SIZE_MILLIS.defaultValue * OperatorContext.APPLICATION_WINDOW_COUNT.defaultValue);
+
+  @FieldSerializer.Bind(JavaSerializer.class)
+  private Duration durationPreventingFreeingSpace;
+
+  private transient StateTracker stateTracker = new StateTracker(this);
+
+  //accessible to StateTracker
+  final transient Object commitLock = new Object();
+
+  @Override
+  public void setup(OperatorContext context)
+  {
+    operatorId = context.getId();
+    fileAccess.init();
+    timeBucketAssigner.register(dataManager);
+    timeBucketAssigner.setup(context);
+
+    numBuckets = getNumBuckets();
+    buckets = new Bucket[numBuckets];
+
+    Preconditions.checkArgument(numReaders <= numBuckets, "no. of readers cannot exceed no. of buckets");
+    //setup state data manager
+    dataManager.setup(context);
+
+    //recovering data for window files to bucket
+    try {
+      Map<Long, Map<Slice, Bucket.BucketedValue>> recovered = dataManager.load(operatorId);
+      if (recovered != null && !recovered.isEmpty()) {
+
+        for (Map.Entry<Long, Map<Slice, Bucket.BucketedValue>> entry : recovered.entrySet()) {
+          int bucketIdx = prepareBucket(entry.getKey());
+
+          for (Map.Entry<Slice, Bucket.BucketedValue> dataEntry : entry.getValue().entrySet()) {
+            buckets[bucketIdx].put(dataEntry.getKey(), dataEntry.getValue().getTimeBucket(),
+                dataEntry.getValue().getValue());
+          }
+        }
+      }
+    } catch (IOException e) {
+      throw new RuntimeException("recovering", e);
+    }
+
+    largestRecoveryWindow = dataManager.getLargestRecoveryWindow();
+    long activationWindow = context.getValue(OperatorContext.ACTIVATION_WINDOW_ID);
+
+    if (activationWindow != Stateless.WINDOW_ID && largestRecoveryWindow <= activationWindow) {
+      replay = true;
+    }
+
+    readerService = Executors.newFixedThreadPool(numReaders, new NameableThreadFactory("managedStateReaders"));
+    stateTracker.setup(context);
+  }
+
+  /**
+   * Gets the number of buckets which is required during setup to create the array of buckets.<br/>
+   * {@link ManagedStateImpl} provides num of buckets which is injected using a property.<br/>
+   * {@link TimeManagedStateImpl} provides num of buckets which are calculated based on time settings.
+   *
+   * @return number of buckets.
+   */
+  public abstract int getNumBuckets();
+
+  public void beginWindow(long l)
```

Comment:
From @amberarrow 
 descriptive parameter name like 'windowId' ? (also in TimeBucketAssigner)


--------
will change